### PR TITLE
Replaced near-cube snapping with grid snapping. Moved Cube to cube.py

### PR DESCRIPTION
**Pull request by djpetti**
Grid snapping is enabled in https://github.com/djpetti/molecube/tree/martin_grid_features.

Still to do:

    Prevent two cubes in same location (swap with location of moved cube?)
    Show grid while dragging

Additionally, when running demo_game, a cube flashing color cannot be properly dragged. If the attempted drag is not the cube that initiated the flashing, the following error occurs:

```
File "C:\Users\Martin\Documents\Programming\molecube\simulator\cube.py", line 252, in clear_drag assert Cube._selected == self```

### CODE REVIEW

File: simulator/cube.py [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+    for side in [Cube.Sides.LEFT, Cube.Sides.RIGHT, Cube.Sides.TOP, Cube.Sides.BOTTOM]:
+      shift = Cube.Sides.coordinates(side)
+
+      # for each cube
+      for other in others:
+        other_x, other_y = other.get_pos()
+        my_x, my_y = self.get_pos() 
+
+        # If other cube is adjacent, add connection
+        if other_x == my_x + shift[0]*grid_size[0] and other_y == my_y + shift[1]*grid_size[1]:
```
Comment (djpetti) [inline simulator/cube.py]: This is very unwieldy. Maybe you should multiply the shift vector by the grid size outside of the loop.

File: simulator/tabletop.py [hunk 2]
```diff
@@ -377,6 +30,7 @@ def __mouse_released(self, event):
       # No cube is selected. Do nothing.
       return
 
+    '''
```
Comment (djpetti) [inline simulator/tabletop.py]: Let's remove the commented-out stuff entirely.

File: simulator/cube.py [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+      side: The side that the sender is connected to.
+      message: The string message being received. """
+    if not self.__application:
+      # With no app, the message gets dropped.
+      return
+
+    # Otherwise, pass it to the app.
+    self.__application.on_message_receive(side, message)
+
+  def is_near(self, other, threshold=100):
```
Comment (djpetti) [inline simulator/cube.py]: I think we should probably get rid of this method since we don't really use it anymore.

File: simulator/tabletop.py [hunk 4]
```diff
@@ -403,6 +57,9 @@ def __mouse_released(self, event):
       # Snap to the closest cube.
       if best_cube:
         selected_cube.snap(best_cube, best_side)
+    '''
+
+    selected_cube.snap_to_grid((200, 200), self.__cubes, offset = (100, 100))
```
Comment (djpetti) [inline simulator/tabletop.py]: I think that the grid size should be defined as a constant. In fact, is there ever a case where we would want the grid size to be different from the cube size? For that matter, you should probably do the same with the offset. Is there any case where the offset will *not* be half the grid size?

File: simulator/cube.py (hunk 3 above)
Comment (duffym4) [reply simulator/cube.py]: I was going to get rid of this method, but remembered you mentioning that the cubes will have the capability to detect nearby but not touching cubes. Is it definitely irrelevant?

File: simulator/cube.py [hunk 5]
```diff
@@ -0,0 +1,401 @@
+import colors
```
[the remark this answers is not in the document]
Comment (duffym4) [reply simulator/cube.py]: Makes sense, will keep this in mind.

File: simulator/word_app.py [hunk 6]
```diff
@@ -123,7 +123,7 @@ class WordGameChecker(Application):
   # Set of valid words.
   _VALID_WORDS = set(["SET", "SEAT", "SEA", "SENT", "NET", "NETS", "TEN",
                       "TENS", "SAT", "TEA", "TEAS", "EAT", "EATS", "NEAT",
-                      "ANT", "ANTS", "ATE"])
+                      "ANT", "ANTS", "ATE", "NEST"])
```
Comment (djpetti) [inline simulator/word_app.py]: haha

File: simulator/tabletop.py [hunk 7]
```diff
@@ -30,37 +34,9 @@ def __mouse_released(self, event):
...
-          best_cube = cube
-
-      # Snap to the closest cube.
-      if best_cube:
-        selected_cube.snap(best_cube, best_side)
-    '''
-
-    selected_cube.snap_to_grid((200, 200), self.__cubes, offset = (100, 100))
-
+    selected_cube.snap_to_grid(Tabletop.GRID_SIZE,
```
Comment (djpetti) [inline simulator/tabletop.py]: Since we're committing to the snapping, it might be cleaner to pass these constants to the `Cube` constructor.

Comment (duffym4) [reply simulator/tabletop.py]: Decided to leave this in Tabletop

**Pull request merged**